### ISSUE TITLE
Timed messages and activities.

### ISSUE THREAD

**Issue by XanderDJ** [aquarial/discord-haskell#73]
I'm trying to implement a bot that helps with running an auction in a discord channel. I figured out how to add state to the bot with the help of an MVar and the eventHandler. But I don't see an entry point where I send messages that say stuff like 5 seconds remaining, and so forth. The only entry point is reacting to events. I could make it so that whenever any event happens the bot checks the time and pushes messages if necessary but this doesn't feel reliable or consistent for holding an auction.

**Comment by aquarial** [comment]
Right analysis of the eventloop. One way to model this is to start each auction "event" in a new thread.


Something like:

```haskell

data Auction = Auction { threadid :: ThreadID, item :: Text }

import UnliftIO.Concurrent (forkIO, threadDelay)

eventHandler :: Event -> DiscordHandler ()
eventHandler event = case event of
      MessageCreate m -> when (shouldStartAuction m) $ do

        -- check in state to see if the user wants to modifiy an existing auction?
        
        
        id <- forkIO $ startAuction
        addAuction (Auction id (messageText m))
        -- add auction to MVar state?


startAuction :: DiscordHandler ()
startAuction = do
   let second = 10^6
   let delay = (second * 60 * 5) # 5 min
   createMessage "Starting auction"
   threadDelay (delay - second * 5)
   createMessage "5 seconds"
   threadDelay (second * 5)
   createMessage "auction closed"
```

**Comment by XanderDJ** [comment]
Sorry for the big delay in responding.

I ended up just having the auctioneer control time management end having the auction happen through commands like intended. This to circumvent this issue. An issue I see with the offered solution is that when someone bids the timer of the auction resets. This to prevent sniping an auction at the last possible second. 

A possible solution I have found for this is using the threadId to kill the thread and start a new one. Data in the mvar would then be used to check if an auction is active or not so that when the thread handling the timer ends, the state gets changed and no bids can occur anymore. 

Thanks for the response!
